openapi-codegen/core/src/main/scala/sttp/tapir/codegen/JsonSerdeGenerator.scala

```diff
@@ -217,8 +171,14 @@ object JsonSerdeGenerator {
   private def generateJsoniterAdtSerde(
       name: String,
       schema: OpenapiSchemaOneOf,
-      maybeFullModelPath: Option[String]
+      maybeFullModelPath: Option[String],
+      jsonParamRefs: Set[String]
   ): String = {
+    val apiTypes = schema.types.collect { case ref: OpenapiSchemaRef if jsonParamRefs.contains(ref.stripped) => ref }
+    if (apiTypes.nonEmpty)
+      throw new NotImplementedError(
+        s"A class cannot be used both in a oneOf at the top level when using jsoniter serdes at $name (problematic children: ${apiTypes})"
```

Comment:
So the problem here is that if there's a serde declared for a subtype of the ADT, then the discriminator is not added on serialization via the parent trait. I don't know how to get around this.